File: app/src/main/java/dev/expositopablo/tonedeath/data/commons/Pinyin.java
package dev.expositopablo.tonedeath.data.commons;

import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.PrimaryKey;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Arrays;

@Entity(tableName = "pinyin_table")
public class Pinyin implements Parcelable {
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    public Long ID = null;

    @NonNull
    @ColumnInfo(name = "Initial")
    public final String Initial;

    @NonNull
    @ColumnInfo(name = "Final")
    public final String Final;

    @NonNull
    @ColumnInfo(name = "ToneVoyelle")
    public final String ToneVoyelle;

    public Pinyin(@NonNull String Initial, @NonNull String Final, @NonNull String ToneVoyelle) {
        this.Initial = Initial;
        this.Final = Final;
        this.ToneVoyelle = ToneVoyelle;
    }

    @Override
    public String toString() {
        return (Initial+Final).replace("∅", "");
    }

    @Override
    public boolean equals(Object obj) {
        if (obj != null && obj instanceof Pinyin){
            Pinyin pinyin = (Pinyin) obj;
            return pinyin.toString().equals(this.toString());
        }
        return false;
    }

    public Long getID() {
        return ID;
    }

    @NonNull
    public String getInitial() {
        return Initial;
    }

    @NonNull
    public String getFinal() {
        return Final;
    }

    @NonNull
    public String getToneVoyelle() {
        return ToneVoyelle;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(this.ID);
        dest.writeString(this.Initial);
        dest.writeString(this.Final);
        dest.writeString(this.ToneVoyelle);
    }

    protected Pinyin(Parcel in) {
        this.ID = (Long) in.readValue(Long.class.getClassLoader());
        this.Initial = in.readString();
        this.Final = in.readString();
        this.ToneVoyelle = in.readString();
    }

    public static final Parcelable.Creator<Pinyin> CREATOR = new Parcelable.Creator<Pinyin>() {
        @Override
        public Pinyin createFromParcel(Parcel source) {
            return new Pinyin(source);
        }

        @Override
        public Pinyin[] newArray(int size) {
            return new Pinyin[size];
        }
    };
}

File: app/src/main/java/dev/expositopablo/tonedeath/data/commons/Pinyin.kt
package dev.expositopablo.tonedeath.data.commons

import android.content.Context
import android.media.AudioManager
import android.media.MediaPlayer
import android.os.Parcelable
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import kotlinx.android.parcel.Parcelize
import java.util.Objects

@Parcelize
@Entity(tableName = "pinyin_table")
data class Pinyin(
        val Initial: String,
        val Final: String,
        val ToneVoyelle: String
): Parcelable {
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    var id: Long? = null

    override fun toString(): String {
        return (Initial + Final)
                .replace("∅", "")
                .replace("ü", "v")
    }

    override fun equals(other: Any?): Boolean {
        if (other != null && other is Pinyin) {
            val pinyin = other as Pinyin?
            return pinyin!!.toString() == this.toString()
        }
        return false
    }

    override fun hashCode(): Int {
        return Objects.hash(this.toString())
    }
}

@Parcelize
data class PinyinDTO(
        val pinyin: Pinyin,
        val tone: String,
        val voice: String
): Parcelable {
    fun getFileName(): String {
        return "${pinyin}_${tone}_$voice"
    }
}

fun PinyinDTO.prepareSoundMediaPlayer(context: Context?, onComplete: (() -> Unit)? = null) : MediaPlayer?{
    return context?.let {
        val resID = it.resources.getIdentifier(this.getFileName(), "raw", it.packageName)
        MediaPlayer.create(it, resID).apply {
            setOnCompletionListener { onComplete?.invoke() }
            setAudioStreamType(AudioManager.STREAM_MUSIC)
        }
    }
}
